lib/mongo_client.js

```diff
@@ -504,6 +505,47 @@ MongoClient.prototype.startSession = function(options) {
   return this.topology.startSession(options, this.s.options);
 };
 
+/**
+ * Runs a given operation with an implicitly created session. The lifetime of the session
+ * will be handled without the need for user interaction.
+ *
+ * NOTE: presently the operation MUST return a Promise (either explicit or implicity as an async function)
+ *
+ * @param {Function} operation An operation to execute with an implicitly created session. The signature of this MUST be `(session) => {}`
+ * @param {Object} [options] Optional settings to be appled to implicitly created session
+ * @return {Promise} returns Promise if no callback passed
+ */
+MongoClient.prototype.withSession = function(operation, options, callback) {
+  if (typeof options === 'function') (callback = options), (options = undefined);
+  const session = this.startSession(options);
+
+  const cleanupHandler = (err, result, opts) => {
```

Comment:
Since `cleanupHandler` should only be called once, we should do:
```js
let cleanupHandler = (err, result, opts) => {
  cleanupHandler = () => {};
  ...
};
```